Odyssey SBE dump collection support code added

Swarnendu-R-C opened the pull request:
We have now OdySsey chip along with the existing p10 which supports SBE dump collection. In future we may have more varities of chips. So the new code to support Odyssey operation has been added.
    
    Test:
    
    busctl --verbose call org.open_power.Dump.Manager /org/openpower/dump xyz.openbmc_project.Dump.Create CreateDump a{sv} 3 "com.ibm.Dump.Create.CreateParameters.DumpType" s "com.ibm.Dump.Create.DumpType.MemoryBufferSBE" "com.ibm.Dump.Create.CreateParameters.ErrorLogId" t 0xDEADBEEF "com.ibm.Dump.Create.CreateParameters.FailingUnitId" t 2
    
    MESSAGE "o" {
    OBJECT_PATH "/xyz/openbmc_project/dump/msbe/entry/40000004";
    };
    
    Jan 20 10:38:42 phosphor-dump-manager[8533]: instanceNum = 2
    Jan 20 10:38:42 phosphor-dump-manager[8533]: Target PATH=/proc0/pib/perv12/mc0/mi0/mcc1/omi0/ocmb0
    Jan 20 10:38:42 phosphor-dump-manager[8533]: Going to call ody_ppe_state
    Jan 20 10:38:42 phosphor-dump-manager[8533]: ppe_get_xir_addr(i_ppe_type, PPE_IDX_XIXCR, i_ppe_instance_num) returned as d0000
    Jan 20 10:38:42 phosphor-dump-manager[8533]: ppe_get_xir_addr(i_ppe_type, PPE_IDX_XIRAMDBG, i_ppe_instance_num) returned as d0003
    Jan 20 10:38:42 phosphor-dump-manager[8533]: ppe_get_xir_addr(i_ppe_type, PPE_IDX_XIRAMEDR, i_ppe_instance_num) returned as d0004
    Jan 20 10:38:42 phosphor-dump-manager[8533]: ppe_get_xir_addr(i_ppe_type, PPE_IDX_XIDBGPRO, i_ppe_instance_num) returned as d0005
    Jan 20 10:38:42 phosphor-dump-manager[8533]: ppe_get_xir_addr(i_ppe_type, PPE_IDX_XIDBGINF, i_ppe_instance_num) returned as d000f
    Jan 20 10:38:43 phosphor-dump-manager[8533]: Done calling ody_ppe_state
    Jan 20 10:38:43 phosphor-dump-manager[8533]: Going to call writeDumpFileForDumpContents
    Jan 20 10:38:43 phosphor-dump-manager[8533]: Entering writeDumpFileForDumpContents
    Jan 20 10:38:43 phosphor-dump-manager[8533]: writng dump data to file
    Jan 20 10:38:43 phosphor-dump-manager[8533]: Exiting writeDumpFileForDumpContents
    Jan 20 10:38:43 phosphor-dump-manager[8533]: Done calling writeDumpFileForDumpContents
    Jan 20 10:38:43 phosphor-dump-manager[8533]: SBE dump collected
    Jan 20 10:38:43 phosphor-dump-manager[8612]: plat_dump/40000004.0_2_SbeData_ody_ody_pibmem_dump
    Jan 20 10:38:43 phosphor-dump-manager[8612]: plat_dump/40000004.0_2_SbeData_ody_ody_pibms_reg_dump
    Jan 20 10:38:43 phosphor-dump-manager[8612]: plat_dump/40000004.0_2_SbeData_ody_ody_ppe_state
    Jan 20 10:38:43 phosphor-dump-manager[8612]: plat_dump/40000004.0_2_SbeData_ody_ody_sbe_localreg_dump
    Jan 20 10:38:43 phosphor-dump-manager[8612]: info.yaml
    Jan 20 10:38:43 phosphor-dump-manager[8521]: Adding Dump Header :/usr/share/dreport.d/include.d/gendumpheader
    Jan 20 10:38:44 bmcweb[899]: (2024-01-20 10:38:44) [ERROR "event_dbus_monitor.hpp":468] Invalid dump type received when listening for dump created signal
    Jan 20 10:38:44 phosphor-dump-manager[8521]: Sat Jan 20 10:38:44 UTC 2024 Successfully completed
    Jan 20 10:38:44 phosphor-dump-manager[496]: OriginatorId is not provided
    Jan 20 10:38:44 phosphor-dump-manager[496]: OriginatorType is not provided. Replacing the string with the default value
    Jan 20 10:38:44 pvm_dump_offload[7879]: Watch interfaceAdded path (/xyz/openbmc_project/dump/bmc/entry/4)
    Jan 20 10:38:44 phosphor-dump-manager[8521]: BMC dump initiated o "/xyz/openbmc_project/dump/bmc/entry/4"
    
    Signed-off-by: Swarnendu Roy Chowdhury <swarnendu.roy.chowdhury@ibm.com>

dhruvibm reviewed (changes requested):
Please align to existing changes..

Swarnendu-R-C commented:
> Please align to existing changes..

Hi, this is aligned with your changes. I just added the code for Odyssey and removed any function which was explicitly made for Odyssey dump collection as per your previous suggestion.

dhruvibm commented on libphal/phal_dump.C:
```diff
@@ -99,37 +104,82 @@ struct DumpPPERegValue {
-	struct pdbg_target* proc = NULL;
-	struct pdbg_target* target;
-	pdbg_for_each_class_target("proc", target)
+	struct pdbg_target* chip = nullptr;
```
change only what is needed for this commit, any improvements should be a separate commit, as I mentioned earlier

Swarnendu-R-C replied:
I have just modified the existing code to accommodate the Odyssey target retrieval  along with P10 target. While doing that I made the necessary changes which is in accordance to modern C++ guidelines. It is not improvement rather slight modification.

dhruvibm commented on libphal/phal_dump.C:
```diff
@@ -99,37 +104,82 @@ struct DumpPPERegValue {
-	struct pdbg_target* target;
-	pdbg_for_each_class_target("proc", target)
+	struct pdbg_target* chip = nullptr;
+	struct pdbg_target* target = nullptr;
```
no change needed in this line

Swarnendu-R-C replied:
As this line has been moved from LN113 to LN140 thus I thought it would be better to change it in accordance to modern C++ practice

dhruvibm replied:
Please keep any improvements in separate commits, as the openbmc codereview guidelines, if the change in number of lines are more we will have to make separate commits, please keep a pull request with minimum change for the odyssey to work

Swarnendu-R-C replied:
If the slightest possible improvements (like changing `NULL` to `nullptr`) needed to be made into different commits then the no of commits would be numerous. Also it is very tedious to maintain the integrity of the changing code in that way. It is my opinion and painful experience and I am ready to discuss elaborately in a meeting or in any respective forum.

dhruvibm commented on libphal/phal_dump.C:
```diff
@@ -99,37 +104,82 @@ struct DumpPPERegValue {
-		if (pdbg_target_index(target) != failingUnit) {
+		uint8_t targetIdx = 0;
+		if (sbeTypeId == ODYSSEY_SBE_DUMP)
+			targetIdx = getFapiUnitPos(target);
```
is FAPI unit position is a unique number across the system?

Swarnendu-R-C replied:
Yes. Earlier without moving ocmb out, the index positions were repeated.. But Fapi positions are unique. As ocmb is not yet moved out, this is the best thing we can do. Confirmed by @deepakala-k

dhruvibm commented on libphal/phal_dump.C:
```diff
@@ -99,37 +104,82 @@ struct DumpPPERegValue {
 			continue;
 		}
-
+		if (sbeTypeId == ODYSSEY_SBE_DUMP) {
```
this looks like redundant error

Swarnendu-R-C replied:
I don't think so as the right target for Odyssey is verified by the function `openpower::phal::sbe::is_ody_ocmb_chip(target)` for an extra layer of security after it comes out from the above statement at LN158

dhruvibm commented on libphal/phal_dump.C:
```diff
@@ -169,473 +219,574 @@ void writeSbeData(const std::filesystem::path& dumpPath, uint32_t reasonCode,
+	if (PROC_SBE_DUMP == sbeTypeId)
+		rv = sbe_get_state(pib_fsi, &state);
+	else if (ODYSSEY_SBE_DUMP == sbeTypeId)
+		rv = sbe_ody_get_state(pib_fsi, &state);
```
can sbe_get_state a common function? both taking same arguments so the differentiation should be inside that function

Swarnendu-R-C replied:
It is a PDBG function which has two different versions for P10 and Odyssey like other HWPs (which takes same types of arguments but performs differently). If @devenrao changes it in PDBG I will also change it here.

dhruvibm commented on libphal/phal_dump.C:
```diff
@@ -169,473 +219,574 @@ void writeSbeData(const std::filesystem::path& dumpPath, uint32_t reasonCode,
 		log(level::ERROR,
-		    "Dump is already collected from the SBE on (%s)",
-		    pdbg_target_path(pib));
+		    "Dump is already collected from the "
```
what is the change in this line?

Swarnendu-R-C replied:
Due to clang format it splitted into two line LN261 & LN262.

dhruvibm commented on libphal/phal_dump.C:
```diff
@@ -169,473 +219,574 @@ void writeSbeData(const std::filesystem::path& dumpPath, uint32_t reasonCode,
  */
-void writeSBERegValuesToFile(const std::vector<DumpSBERegVal>& dumpRegs,
-			     const std::filesystem::path& basePath)
+void extractSbeRc(struct pdbg_target* target,
```
why this function name changed?

dhruvibm replied:
you may change the name as appropriate but not in this commit

Swarnendu-R-C replied:
This is a different function than `writeSBERegValuesToFile`. I didn't get you here.

dhruvibm replied:
extractSbeRc, why this changed?

Swarnendu-R-C replied:
It was `executeExtractSbeRc` I guess. I removed "execute" because it sounds redundant (as any function call is made to execute the logic within it). Open to discuss it further.

dhruvibm commented on libphal/phal_dump.C:
```diff
@@ -169,473 +219,574 @@ void writeSbeData(const std::filesystem::path& dumpPath, uint32_t reasonCode,
-void collectPIBMSRegDump(struct pdbg_target* proc,
-			 const std::filesystem::path& dumpPath,
-			 const std::string& baseFilename)
+struct pdbg_target* preCollection(const uint32_t failingUnit,
```
this function was removed

Swarnendu-R-C replied:
No, it is just been moved at LN604

dhruvibm commented on libphal/phal_dump.C:
```diff
@@ -169,473 +219,574 @@ void writeSbeData(const std::filesystem::path& dumpPath, uint32_t reasonCode,
  */
-void writePIBMEMDataToFile(const std::vector<uint64_t>& dumpData,
-			   const std::filesystem::path& basePath)
+void writeDumpFile(char* data, size_t len, std::filesystem::path& dumpPath)
```
this function was removed

Swarnendu-R-C replied:
Rather than having four different functions for writing four different HWPs dump contents with same lines of code and logic I have tried to optimize it into a single function which will handle the dump contents from the all four HWPs, thus removed.

dhruvibm replied:
optimization in a different pull request with a different story

Swarnendu-R-C replied:
This function `writeDumpFile` is not been changed actually. Only it's place has been changed due to introduction of other functions above.

dhruvibm commented on libphal/phal_dump.C:
```diff
@@ -169,473 +219,574 @@ void writeSbeData(const std::filesystem::path& dumpPath, uint32_t reasonCode,
  */
-void writePPEStateToFile(const std::vector<DumpPPERegValue>& ppeState,
-			 const std::filesystem::path& basePath)
+void collectPPEStateData(struct pdbg_target* target,
```
why. the order of functons changed?

Swarnendu-R-C replied:
I tried to declare and define the functions as near as possible of theirs usages and in group for easy readability and standard

dhruvibm replied:
any such changes with a different pull request, keep the commit to the scope of the change

Swarnendu-R-C replied:
Will be discussing it in a meeting perhaps.

dhruvibm commented on libphal/phal_dump.C:
```diff
@@ -169,473 +219,574 @@ void writeSbeData(const std::filesystem::path& dumpPath, uint32_t reasonCode,
-		       const std::filesystem::path& dumpPath,
-		       const std::string& baseFilename)
+template <typename DumpRegVal>
+void writeDumpFileForDumpContents(std::vector<DumpRegVal>& dumpRegs,
```
why this new function added?

Swarnendu-R-C replied:
Rather than having four different functions for writing four different HWPs dump contents with same lines of code and logic I have tried to optimize it into a single function by making a templated one which will handle the dump contents from the all four HWPs.

Swarnendu-R-C replied:
Again my point is if I am touching some sections of the existing code during a major enhancement process wouldn't it be better to get it tidy up as much as possible in one go? I am ready to discuss it further in a meeting perhaps.

dhruvibm reviewed (changes requested):
I see a lot of changes which may not be necessary to get the Odyssy SBE dump working in this commit, so please make just the changes needs to get the SBE dump working in this commit and do any refactoring or improvements in a separate commits..

Swarnendu-R-C reviewed (commented):
I have had changed the code on top of the latest restructured code as I seemed deemed to fit and appropriate from optimization point of view. It is very hard to segregate the Odyssey code from P10 and splitting into different commits as the changes are merely if-else conditions with either same logic or different HWP/PDBG function calls (on previous suggestions I have removed the dedicated methods for Odyssey and integrated into the existing P10 one totally). That would also lead to high chances of missing the necessary pieces of code, prone to typos and retesting multiple times.  I am ready for an elaborate discussion if I sound confusing anywhere or misunderstanding anything.

dhruvibm commented:
Please keep the changes only to the scope of this pull request, any optimizations with well-defined commit messages and test cases.

Swarnendu-R-C reviewed (commented):
I am really puzzled to understand that who writes the code in an un-optimized way first, then optimized it and then beautify it in separate instances when s/he knows all these three aspects (what exactly needs to be done) of the changing code beforehand? It is tedious, time taking, requires same testing multiple times (which is time consuming in cases like SBE dump collection for Odyssey). Perhaps I need a meeting to understand this. Also if needed be I am ready to bring up my doubts in any respective forum or discussion. Perhaps the guidelines need a little bit tweeks or may be we are mis interpreting it somehow.

Swarnendu-R-C commented:
Closing this PR as the conversation has gone quite a long one. Created a new PR https://github.com/open-power/ipl/pull/78 where all the suggestions are applied.
